### DISCUSSION TITLE
Episode 126: Flux & Flow with Tristan Brindle

### DISCUSSION THREAD

**Discussion by codereport** [codereport/adsp2#17] [Episode]
Episode Link: https://adspthepodcast.com/2023/04/21/Episode-126.html

Feel free to write comments or ask questions 🙂

**Comment by tylerjw** [comment]
I take a slight issue with Barry's talk about Rust iterators in that he talks about them being based on the single function `next` and that there is no way to advance without dereferencing. To make a type that is iterable in rust the minimal interface is `next`. There are functions that are by default implemented in the `Iterator` trait like `nth` to advance further than just the next iteration. Many iterable types (like the iterator type you can get from a `slice`) specialize `nth` so it is advanced without dereferencing.

Barry states that the Rust model is simpler and lacks functionality in the C++ model for composable iteration. Rust makes it simpler to implement a type that can be treated as an iterable but it is by no means lacking this small bit of functionality.

Also, Rust iterators have a straightforward interface to express interior immutability or consuming iteration. These two bits of functionality seem to be missing from the C++ implementation.

~~I wonder if Barry made this mistake because he didn't read the source for Rust Iterarors but instead based his talk on someone's C++ library, which used only the next function.~~ unhelpful speculation

![image](https://user-images.githubusercontent.com/2490389/234318021-c75463b7-de8d-4c77-981d-bed293a0f9b2.png)

**Reply by codereport** [reply]
@tylerjw Thanks for the comment Tyler. Your message is mostly constructive but the last paragraph I think is entirely unnecessary. Speculating on why a mistake was or wasn't made isn't really helpful. And calling something "silly" is never necessary. You are basically implying he didn't do the appropriate amount of research which is unkind.

**Reply by tylerjw** [reply]
I found Barry's talk helpful in understanding the fine details of how C++ ranges library works, even if I found his conclusion unfair and uninformed. It is fair to point this out because his point about Rust having a less-functional iteration model is a core part of this talk and not some small mistake.

**Reply by tylerjw** [reply]
I updated my comment to remove the statement about it being a silly mistake. I realize we are all doing our best to understand these things, and preparing a talk like this is a lot of work, and making this mistake is understandable. I found it frustrating to watch this talk because this mistake greatly influences the core point of this talk.

**Reply by tcbrindle** [reply]
Just realised that I accidentally added a new top-level comment rather than replying in-line, so please [click here](https://github.com/codereport/adsp2/discussions/17#discussioncomment-5767121) for my response in case you don't get a ping from Github.

**Comment by tcbrindle** [comment]
>  It is fair to point this out because his point about Rust having a less-functional iteration model is a core part of this talk and not some small mistake.

It's not a mistake, and @brevzin's conclusions about the Rust iterator model are correct. A key observation is this: there is no `sort()` method in the Rust standard library that operates on iterators -- rather, Rust's `sort()` operates on slices (equivalent to `std::span` in C++).

What about `nth()`, then? From the [Rust documentation](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.nth), we can see that this is specified to effectively behave as if implemented with repeated calls to `next()`. While some iterator implementations are able to do it more efficiently, there is no iterator extension trait that guarantees this behavior, and so in generic code we need to assume that `nth()` is always an `O(N)` operation.

But even if we did introduce an `IteratorWithConstantTimeNth` trait, it still wouldn't be sufficient to do all the things a C++ random access iterator can do, nor all the things that random-access algorithms like `sort()` require. We would also need, in no particular order:

 * Some way of repeatedly accessing the current element without incrementing the iterator (like `peek()`, but returns the element we're at *now*, not the next one)
 * Some way of moving *backwards* from the current position an arbitrary number of places in constant time: `nth()` takes an unsigned value and can only move forward, and is documented as "consuming" all the intermediate values: we'd need some way of "unconsuming" them again, which Rust iterators don't support
 * An O(1) way of comparing two iterators to decide if one is at an earlier or later position in the sequence than the other (C++ RA iterators do this by overloading the usual comparison operators)
 * A way of accessing, in constant time, the distance between two iterators: that is, what argument would we need to give to `nth(x)` to move one iterator to the same sequence position as the second?

In other words, it's not just about making `nth()` constant time: we'd need a lot of other stuff as well to get equivalent functionality to C++, which would significantly alter (and complicate) the Rust iterator design.

If you think about how you might do these things, and do them in user-friendly way with a nice API, and still obey the Rust rules about only a single mutable borrow at a time, you'll probably end up with a design pretty similar to the [Swift `RandomAccessCollection` protocol](https://developer.apple.com/documentation/swift/randomaccesscollection) -- which (coincidentally) I also ended up doing in Flux.

But it can't be done with Rust standard library iterators today, which is the point Barry was making in his talk.

**Reply by tylerjw** [reply]
Thank you for the detailed and thoughtful response. I was wrong and learned something constructive through this interaction. Thank you for your work on this.
